cmd/tests/tracing_module_test.go

```diff
@@ -99,6 +100,107 @@ func TestTracingClient_DoesNotInterfereWithHTTPModule(t *testing.T) {
 	assert.Equal(t, int64(2), atomic.LoadInt64(&gotInstrumentedRequests))
 }
 
+func TestTracingModuleClient_HundredPercentSampling(t *testing.T) {
+	t.Parallel()
+	tb := httpmultibin.NewHTTPMultiBin(t)
+
+	var gotRequests int64
+	var gotSampleFlags int64
+
+	tb.Mux.HandleFunc("/tracing", func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt64(&gotRequests, 1)
+
+		traceparent := r.Header.Get("traceparent")
+		require.NotEmpty(t, traceparent)
+		require.Len(t, traceparent, 55)
+
+		if traceparent[54] == '1' {
+			atomic.AddInt64(&gotSampleFlags, 1)
+		}
+	})
+
+	script := tb.Replacer.Replace(`
+		import http from "k6/http";
+		import { check } from "k6";
+		import tracing from "k6/experimental/tracing";
+
+		export const options = {
+			// 100 iterations to make sure we get 100% sampling
+			iterations: 100,
+		}
+
+		const instrumentedHTTP = new tracing.Client({
+			propagator: "w3c",
+
+			// 100% sampling
+			sampling: 100,
+		})
+
+		export default function () {
+			instrumentedHTTP.get("HTTPBIN_IP_URL/tracing");
+		};
+	`)
+
+	ts := getSingleFileTestState(t, script, []string{"--out", "json=results.json"}, 0)
+	cmd.ExecuteWithGlobalState(ts.GlobalState)
+
+	assert.Equal(t, int64(100), atomic.LoadInt64(&gotSampleFlags))
+	assert.Equal(t, int64(100), atomic.LoadInt64(&gotRequests))
+
+	jsonResults, err := afero.ReadFile(ts.FS, "results.json")
+	require.NoError(t, err)
+
+	assertHasTraceIDMetadata(t, jsonResults, 100, tb.Replacer.Replace("HTTPBIN_IP_URL/tracing"))
+}
+
+func TestTracingModuleClient_ZeroPercentSampling(t *testing.T) {
+	t.Parallel()
+	tb := httpmultibin.NewHTTPMultiBin(t)
+
+	var gotRequests int64
+	var gotSampleFlags int64
+
+	tb.Mux.HandleFunc("/tracing", func(w http.ResponseWriter, r *http.Request) {
+		atomic.AddInt64(&gotRequests, 1)
+
+		traceparent := r.Header.Get("traceparent")
+		require.NotEmpty(t, traceparent)
+		require.Len(t, traceparent, 55)
+
+		if traceparent[54] == '1' {
+			atomic.AddInt64(&gotSampleFlags, 1)
+		}
+	})
+
+	script := tb.Replacer.Replace(`
+		import http from "k6/http";
+		import { check } from "k6";
+		import tracing from "k6/experimental/tracing";
+
+		export const options = {
+			// 100 iterations to make sure we get 100% sampling
+			iterations: 100,
+		}
+
+		const instrumentedHTTP = new tracing.Client({
+			propagator: "w3c",
+
+			// 100% sampling
+			sampling: 0,
```

Comment:
Can we maybe have a test with sampling of like `10` as that is more likely wht users will do - sample some non-0 non-100 percent.

Reply:
The reason I did not have such a test is that using the current probabilistic sampling technique the only insurance we have is that on a large enough number of samples, we will be converging towards 10% of sampled operations.

Maybe that's a good enough reason to consider switching to a more deterministic sampling strategy 🤔 I was thinking an alternative might be filling a boolean buffer of a fixed size with P% `true` values, and shuffling it every time we observed all of its values. That way we can assert that for a given buffer/window size we indeed have P% sampled operations.

What do you think?

Reply:
As I'm at it, can I get your input on this @Blinkuu, I imagine you might have valuable input on this ? cheers 🤓

Reply:
Sorry, I'm not entirely sure what you mean here. How would it differ if you used a slice of `bool` and traversed it with a pointer instead of making a decision on-the-fly in terms of determinicity? Could you elaborate on both ideas?

Reply:
It would differ in that with the slice of `bool` having a fixed size (that I would refer to as "window" size), I would have some capacity to test the outcome of sampling. As I would know the window size ("the slice contains 256 booleans, 10% of which should be `false` before we re-shuffle it"), I can, in a test, check that it contains the expected percentage of positives before we reshuffle it.

Whereas with the current probabilistic sampling strategy, as far as I found out through some research, it wouldn't be relevant to try to test that P% of the samples are picked, as every "pick" has a chance of P% but is random. For a 10% sampling rater, I could test that over 100K iterations, the number of picked samples would converge towards 10K, but there is still a (small) chance to miss that by a large shot and be whatever (7K); leading to a flaky test.

Reply:
> I can, in a test, check that it contains the expected percentage of positives before we reshuffle it.

I believe there are better ways to test code that depends on randomness. You can simply accept an interface that generates random values and mock it in the tests for determinism. You can look at how we do it in the insights project [here](https://github.com/grafana/k6-cloud-insights/blob/4bf1ac687a489568848b7e7541e549ec07379aa4/pkg/util/traceid/traceid_test.go#L141).

Regarding testing you have "exactly" 10% of requests being sampled - I see little value in such a test, especially if it were to be a "unit test".

Reply:
Coincidentally, I added support for custom randomness generation in another PR #2894 already. I guess that's a good data point indicating we want to actually do it then. I'll implement it in this PR too then. Thanks a lot for your input 🙇🏻

Reply:
As testing a random behavior makes little sense, and I believe testing a mocked randomness behavior wouldn't bring much value either. I'll close this. Feel free to reopen if you have a concrete use case and approach for testing this reliably 👍🏻